quiche/src/lib.rs

```diff
@@ -4374,13 +4374,14 @@ impl Connection {
             self.blocked_limit = Some(self.max_tx_data);
         }
 
+        let cap = self.tx_cap;
+
         // Truncate the input buffer based on the connection's send capacity if
         // necessary.
         //
         // When the cap is zero, the method returns Ok(0) *only* when the passed
         // buffer is empty. We return Error::Done otherwise.
-        let cap = self.tx_cap;
-        if cap == 0 && !(fin && buf.is_empty()) {
+        if cap == 0 && !buf.is_empty() {
```

Comment:
Why is this line changed? It would seem to change behaviour compared to what the docs say.

Reply:
There is no actual behaviour change from the application POV. Also, the documentation talks about "0-length" buffers, not "0-length and final" ones when returning `Ok(0)` instead of `Done`, so it seems like this aligns better with that anyway.

In any case this is required to make sure the state of the stream is updated properly when doing 0-length writes (e.g. https://github.com/cloudflare/quiche/blob/master/quiche/src/h3/mod.rs#L1009), otherwise it doesn't get woken up later.

Reply:
That is, on `stream_send()` calls with 0-length buffers this makes it go a bit further in the code rather than return early, but the end result is the same.

Reply:
To add more details, the stream state wouldn't actually be created when doing a 0-length write because we would return early, so the code in the h3 code was useless. It wasn't a problem before, but now we actually need the state so that the stream can be marked as writable when it gets more capacity.

Reply:
cheers for the explanation, makes sense